daemon/MenuDaemon.vala

```diff
@@ -0,0 +1,189 @@
+//
+//  Copyright (c) 2018 elementary LLC. (https://elementary.io)
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+namespace GalaDaemon {
+	[DBus (name = "org.pantheon.gala")]
+	public interface GalaInterface : GLib.Object
+	{
+		public abstract void perform_action (Gala.ActionType type);
+	}
+
+	[DBus (name = "io.elementary.gala.daemon")]
+	public class MenuDaemon : Object
+	{
+		private const string MENU_DBUS_NAME = "io.elementary.gala.daemon";
+		private const string MENU_DBUS_OBJECT_PATH = "/io/elementary/gala/daemon";
+
+		private const string GALA_DBUS_NAME = "org.pantheon.gala";
+		private const string GALA_DBUS_OBJECT_PATH = "/org/pantheon/gala";
+
+		private Gtk.Menu? window_menu = null;
+		private Gtk.MenuItem minimize;
+		private Gtk.MenuItem maximize;
+		private Gtk.MenuItem move;
+		private Gtk.MenuItem resize;
+		private Gtk.CheckMenuItem always_on_top;
+		private Gtk.CheckMenuItem on_visible_workspace;
+		private Gtk.MenuItem move_left;
+		private Gtk.MenuItem move_right;
+		private Gtk.MenuItem close;
+
+		private GalaInterface? gala_proxy = null;
+
+		[DBus (visible = false)]
+		public void setup_dbus ()
+		{
+			var flags = BusNameOwnerFlags.ALLOW_REPLACEMENT | BusNameOwnerFlags.REPLACE;
+			Bus.own_name (BusType.SESSION, MENU_DBUS_NAME, flags, on_bus_acquired, () => {}, null);
+
+			Bus.watch_name (BusType.SESSION, GALA_DBUS_NAME, BusNameWatcherFlags.NONE, has_gala, lost_gala);
+		}
+
+		void on_gala_get (GLib.Object? o, GLib.AsyncResult? res)
+		{
+			try {
+				gala_proxy = Bus.get_proxy.end (res);
+			} catch (Error e) {
+				warning ("Failed to get Gala proxy: %s", e.message);
+			}
+		}
+
+		void lost_gala ()
+		{
+			gala_proxy = null;
+		}
+
+		void has_gala ()
```

Comment:
I would prefer this method to be named something like `gala_appeared`, judging by the name, I would think it returns if there is a gala connection currently established instead of actually connecting to it.